dss/storage/index_document.py

```diff
@@ -85,17 +127,81 @@ def files(self):
+        if tombstone is None:
+            index_name = self._prepare_index(dryrun)
+            return self._index_into(index_name, dryrun)
+        else:
+            self.logger.info(f"Found tombstone for {self.fqid}. Indexing tombstone in place of bundle.")
+            return self.entomb(tombstone)
+
+    def _lookup_tombstone(self):
+        for all_versions in (False, True):
+            tombstone_id = self.fqid.to_tombstone_id(all_versions=all_versions)
+            try:
+                return BundleTombstoneDocument.from_replica(self.replica, tombstone_id, self.logger)
+            except BlobNotFoundError:
+                pass
+        return None
+
+    def _index_into(self, index_name: str, dryrun: bool):
+        elasticsearch_retry.add_context(index=index_name)
+        msg = self._msg(dryrun)
+        versions = self._get_indexed_versions()
+        old_version = versions.pop(index_name, None)
+        if versions:
+            self.logger.warning(msg(f"Removing stale copies of the bundle document for {self.fqid} from the following "
+                                    f"index(es): {json.dumps(versions)}."))
+            if not dryrun:
+                self._remove_versions(versions)
+        if old_version:
+            old_doc = self.from_index(self.replica, self.fqid, index_name, self.logger, version=old_version)
+            if self == old_doc:
+                self.logger.info(f"Document for bundle {self.fqid} is already up-to-date in index {index_name} at "
+                                 f"version {old_version}.")
+                return False, index_name
+            else:
+                self.logger.warning(msg(f"Updating an older copy of the document for bundle {self.fqid} in index "
+                                        f"{index_name} at version {old_version}."))
+        else:
+            self.logger.info(msg(f"Writing the document for bundle {self.fqid} to index "
+                                 f"{index_name} for the first time."))
+        if not dryrun:
+            self._write_to_index(index_name, version=old_version or 0)
+        return True, index_name
+
+    @elasticsearch_retry
+    def entomb(self, tombstone: 'BundleTombstoneDocument', dryrun=False):
+        """
+        Ensure that there is exactly one up-to-date instance of a tombstone for this document in exactly one
+        ES index. The tombstone data overrides the document's data in the index.
+
+        :param tombstone: The document with which to replace this document in the index.
+        :param dryrun: see :py:meth:`~IndexDocument.index`
+        :return: see :py:meth:`~IndexDocument.index`
+        """
+        elasticsearch_retry.add_context(bundle=self, tombstone=tombstone)
+        self.logger.info(f"Writing tombstone for {self.replica.name} bundle: {self.fqid}")
+        # Preare the index using the original data such that the tombstone can be placed in the correct index.
+        index_name = self._prepare_index(dryrun)
```

Comment:
As implemented here, indexing a bundle that has been admin deleted will fail if the bundle data has been deleted from the bucket. This will "work" in the short-term, as admin delete does not actually delete the bundle data. Yet, going forward, it would be best if indexing of an admin deleted bundle did not depend on the bundle data existing in the bucket!

Reply:
https://github.com/HumanCellAtlas/data-store/issues/821